init.c

```diff
@@ -143,6 +143,117 @@ static int other_mounts()
 	return 0;
 }
 
+struct pv_init_daemon {
+	char *name;
+	pid_t pid;
+	int respawn;
+	char *testpath;
+	char *cmd;
+	int _respawning;
+};
+
+struct pv_init_daemon daemons[] = { { "hwrngd", 0, 1, "/usr/sbin/rngd",
+				      "/usr/sbin/rngd -f", 0 },
+				    { 0, 0, 0, 0, 0, 0 } };
+
+static int daemon_spawn(struct pv_init_daemon *self)
+{
+	self->pid = 0;
+	pv_log(INFO, "Spawning rngd daemon.");
+	if (self->_respawning) {
+		pv_log(INFO, "... deferring respawn by 5 seconds");
+		self->pid = tsh_run("/bin/sleep 5", 0, 0);
+		self->_respawning = 0;
+	} else {
+		self->pid = tsh_run(self->cmd, 0, 0);
+	}
+	if (self->pid < 0) {
+		pv_log(ERROR, "error forking child: '%s': %s", self->name,
+		       strerror(errno));
+		return self->pid;
+	}
+
+	return self->pid;
+}
+
+static int pv_init_spawn_daemons()
+{
+	int i = 0;
+	struct stat sb;
+
+	sigset_t blocked_sig, old_sigset;
+	sigemptyset(&blocked_sig);
+	sigaddset(&blocked_sig, SIGCHLD);
+	sigprocmask(SIG_BLOCK, &blocked_sig, &old_sigset);
+
+	while (daemons[i].name) {
+		if (daemons[i].pid > 0 ||
+		    (daemons[i].pid == 0 && !daemons[i].respawn))
+			goto next;
+
+		if (stat(daemons[i].testpath, &sb)) {
+			pv_log(INFO,
+			       "daemon not enabled %s: disabling respawn\n",
+			       daemons[i].name);
+			daemons[i].pid = 0;
+			daemons[i].respawn = 0;
+			goto next;
+		}
+
+		if (!daemons[i]._respawning)
+			daemons[i]._respawning = 1;
+
+		daemons[i].pid = daemon_spawn(daemons);
+		pv_log(INFO, "spawned daemon %s: %d \n", daemons[i].name,
+		       daemons[i].pid);
+	next:
+		i++;
+	}
```

Comment:
I always tried to avoid "goto" so:
```suggestion
	for (int i = 0; daemons[i].name; ++i) {
		if (daemons[i].pid > 0 ||
		    (daemons[i].pid == 0 && !daemons[i].respawn))
			continue;

		if (stat(daemons[i].testpath, &sb)) {
			pv_log(INFO,
			       "daemon not enabled %s: disabling respawn\n",
			       daemons[i].name);
			daemons[i].pid = 0;
			daemons[i].respawn = 0;
			continue;
		}

		if (!daemons[i]._respawning)
			daemons[i]._respawning = 1;

		daemons[i].pid = daemon_spawn(daemons);
		pv_log(INFO, "spawned daemon %s: %d \n", daemons[i].name,
		       daemons[i].pid);
	}
```
Maybe there is no huge difference, but it feels better (at least to me)

Reply:
ok but why ++i instead of i++ ?

Reply:
it's just habit

Reply:
did that in latest rev.